src/v/archival/ntp_archiver_service.cc

```diff
@@ -1823,6 +1833,248 @@ ss::future<> ntp_archiver::housekeeping() {
+            std::swap(tmp, rem_batch);
+            if (co_await batch_delete(std::move(tmp))) {
+                successful_deletes += sz;
+            }
+        }
+    }
+    if (co_await batch_delete(rem_batch)) {
+        successful_deletes += rem_batch.size();
+    }
+    rem_batch.clear();
+
+    const auto backlog_size_exceeded = segments_to_remove.size()
+                                       > _max_segments_pending_deletion();
+    const auto all_deletes_succeeded = successful_deletes
+                                       == segments_to_remove.size();
+
+    if (!all_deletes_succeeded && backlog_size_exceeded) {
+        vlog(
+          _rtclog.warn,
+          "The current number of spillover segments pending deletion has "
+          "exceeded the configurable limit ({} > {}) and deletion of some "
+          "segments failed. Metadata for all remaining segments pending "
+          "deletion will be removed and these segments will have to be removed "
+          "manually.",
+          segments_to_remove.size(),
+          _max_segments_pending_deletion());
+    }
+    if (!all_deletes_succeeded && !backlog_size_exceeded) {
+        vlog(
+          _rtclog.info,
+          "Failed to delete all selected segments from cloud storage. Will "
+          "retry on the next housekeeping run.");
+    }
+    if (all_deletes_succeeded || backlog_size_exceeded) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _parent.archival_meta_stm()->cleanup_archive(
+          new_clean_offset, bytes_to_remove, deadline, _as);
+
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.info,
+              "Failed to clean up metadata after garbage collection: {}",
+              error);
+        } else {
+            // Remove manifests only if metadata no longer references them
+            for (const auto& path : manifests_to_remove) {
+                rem_batch.emplace_back(path);
+                if (rem_batch.size() >= batch_size) {
+                    std::vector<cloud_storage_clients::object_key> tmp;
+                    std::swap(tmp, rem_batch);
+                    co_await batch_delete(std::move(tmp));
+                }
+            }
+            co_await batch_delete(rem_batch);
+        }
+    }
+    _probe->segments_deleted(static_cast<int64_t>(successful_deletes));
+    vlog(
+      _rtclog.debug,
+      "Deleted {} spillover segments from the cloud",
+      successful_deletes);
+}
+
+ss::future<bool> ntp_archiver::batch_delete(
+  std::vector<cloud_storage_clients::object_key> keys) {
+    if (_remote.is_batch_delete_supported()) {
+        // Do batch delete, the batch size should be below the limit
+        auto res = co_await _remote.delete_objects(
+          get_bucket_name(), std::move(keys), _rtcnode);
```

Comment:
`remote::delete_objects` does not currently consider the request failed if some keys were not deleted. This will trick the code in `garbage_collect_archive` into thinking that it always deleted everything. Maybe cherry pick [this commit](https://github.com/VladLazar/redpanda/commit/1c288019615fda1e196fa58206c8566180ef1ff4) that fixes `delete_objects` semantics.